### PR TITLE
clean up metrics code, use MetricCollection

### PR DESCRIPTION
**Pull request by adamgayoso**
By calling .update() explicitly, it should speed things up according to their documentation as we don't need the per step values.

This also fixes an issue where `n_obs_training` might not reflect the number of cells in an epoch, as in the large scale training params where we restrict an epoch to 20 minibatches

### CODE REVIEW

File: scvi/train/_metrics.py [hunk 1]
```diff
@@ -45,29 +39,115 @@ def mode(self):
...
+
+    def update(
+        self,
+        kl_global: torch.Tensor,
+        n_obs_minibatch: int,
+    ):
+        """Updates all metrics."""
+        super().update(kl_global, n_obs_minibatch)
+
+    def compute(self):
```
Comment (jjhong922) [inline scvi/train/_metrics.py:150]: this looks like the only thing that sets apart this class from the rest. Otherwise a lot of copy paste. To avoid over generalizing, maybe you could just stick to having the base class, except it takes a metric name, and you add the kwarg for what to normalize by?

File: scvi/train/_trainingplans.py [hunk 2]
```diff
@@ -136,14 +137,49 @@ def __init__(
...
+            self.rec_loss_val,
+            self.kl_local_val,
+            self.kl_global_val,
+            self.val_metrics,
+        ) = self._create_elbo_metric_components(
+            mode="validation", n_total=self.n_obs_validation
+        )
         self.elbo_val.reset()
 
+    def _create_elbo_metric_components(self, mode: str, n_total: Optional[int] = None):
```
Comment (jjhong922) [inline scvi/train/_trainingplans.py:164]: nit: move to top of class (before usage)

File: scvi/train/_trainingplans.py [hunk 3]
```diff
@@ -136,14 +137,49 @@ def __init__(
...
+        """Initialize ELBO metric and the metric collection."""
+        rec_loss = ReconstructionLossMetric(mode)
```
Comment (jjhong922) [inline scvi/train/_trainingplans.py:166]: based on my first comment this could be replaced by `ElboMetric("reconstruction_loss", mode, "n_obs")` etc.

File: scvi/train/_trainingplans.py [hunk 4]
```diff
@@ -136,14 +137,49 @@ def __init__(
...
+        kl_local = KLLocalMetric(mode)
+        kl_global = KLGlobalMetric(mode)
+        # n_total can be 0 if there is no validation set, this won't ever be used
```
Comment (jjhong922) [inline scvi/train/_trainingplans.py:169]: can add an assertion statement here for this? `raise AssertionError()` to satisfy codacy

**Comment by jjhong922** [comment]
@PierreBoyeau  @adamgayoso  updated and addressed my own comments. Condensed code by relying on this part of the metricscollection code https://github.com/PyTorchLightning/metrics/blob/11641b8558e713649743709d78d3ab363bcfc9aa/torchmetrics/metric.py#L681-L682

**Comment by jjhong922** [comment]
Also noticed GIMVI doesnt use this and seems hard to adapt to it. Does it matter?

File: scvi/train/_trainingplans.py [hunk 5]
```diff
@@ -209,44 +245,15 @@ def compute_and_log_metrics(
...
-            reduce_fx=torch.sum,
-            on_step=False,
-            on_epoch=True,
-            batch_size=n_obs_minibatch,
-        )
-        # default aggregation is mean
-        self.log(
-            f"kl_global_{mode}",
-            kl_global,
+        self.log_dict(
```
Comment (PierreBoyeau) [inline scvi/train/_trainingplans.py:255]: One naive question as this is not fresh in my mind anymore. Where are the metrics explicitly reset to 0?

Comment (jjhong922) [reply scvi/train/_trainingplans.py:256]: They get set to 0 on __init__ https://github.com/scverse/scvi-tools/pull/1529/files/2eb5cb855bfaa3fc2a7a45f057928fe1742dabc6#diff-8fc39d9964444c66caf3545a773bef7402a7a73386aa2761f06930d7759353daR135 as well as when `n_obs_training` is set, which happens in https://github.com/scverse/scvi-tools/blob/master/scvi/train/_trainrunner.py#L70.

File: scvi/train/_metrics.py [hunk 6]
```diff
@@ -10,64 +10,81 @@ class ElboMetric(Metric):
...
+
     def update(
         self,
-        reconstruction_loss_sum: torch.Tensor,
-        kl_local_sum: torch.Tensor,
-        kl_global: torch.Tensor,
-        n_obs_minibatch: int,
+        **kwargs,
     ):
         """Updates all metrics."""
```
Comment (adamgayoso) [inline scvi/train/_metrics.py:74]: might be good to explain somehwere that in a metric collection that arguments get passed to each individual metric
